src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -129,6 +134,60 @@ split_change_cb(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
 	}
 
 	Relation targetRelation = RelationIdGetRelation(targetRelationOid);
+
+	/*
+	 * If any columns from source relation have been dropped, then the tuple needs to
+	 * be formatted according to the target relation.
+	 */
+	TupleDesc sourceRelationDesc = RelationGetDescr(relation);
+	TupleDesc targetRelationDesc = RelationGetDescr(targetRelation);
+	if (sourceRelationDesc->natts > targetRelationDesc->natts)
+	{
+		HeapTuple targetRelationTuple = GetTupleForTargetSchema(existingTuple,
+																sourceRelationDesc,
+																targetRelationDesc);
+		switch (change->action)
+		{
+			case REORDER_BUFFER_CHANGE_INSERT:
+			{
+				change->data.tp.newtuple->tuple = *targetRelationTuple;
+				break;
+			}
+
+			case REORDER_BUFFER_CHANGE_UPDATE:
+			{
+				change->data.tp.newtuple->tuple = *targetRelationTuple;
+
+				/*
+				 * For a relation with REPLICA_IDENTITY_FULL, we need to format the old tuple too
+				 * according to the target relation.
+				 */
+				if (relation->rd_rel->relreplident == REPLICA_IDENTITY_FULL)
```

Comment:
we also need to do this for delete, I think

Reply:
In DELETE case, we are always reformatting the tuple.

Reply:
> In DELETE case, we are always reformatting the oldtuple.

hmz, interesting, can you share a link to the postgres code?

Reply:
> > In DELETE case, we are always reformatting the oldtuple.
> 
> hmz, interesting, can you share a link to the postgres code?

I meant in my current PR code.

Reply:
Ah, right. I had realized this, but then forgot when I briefly looked at it again. I think my confusion comes from the fact that we treat oldtuple differently in case of update and delete. Should we skip in case of a not full replica identity for delete as well?